qiskit/visualization/utils.py

```diff
@@ -96,88 +98,34 @@ def _get_layered_instructions(circuit, reverse_bits=False, justify=None, idle_wi
             ops.append([node])
 
     if justify == 'left':
-        for dag_layer in dag.layers():
-            layers = []
-            current_layer = []
-
-            dag_nodes = dag_layer['graph'].op_nodes()
-            dag_nodes.sort(key=lambda nd: nd._node_id)
+        spooler = LayerSpooler(qregs, Justification.LEFT)
 
+        for dag_layer in dag.layers():
+            current_index = spooler.last_index()
+            dag_nodes = _sorted_nodes(dag_layer)
             for node in dag_nodes:
-                multibit_gate = len(node.qargs) + len(node.cargs) > 1
-
-                if multibit_gate:
-                    # need to see if it crosses over any other nodes
-                    gate_span = _get_gate_span(qregs, node)
-
-                    all_indices = []
-                    for check_node in dag_nodes:
-                        if check_node != node:
-                            all_indices += _get_gate_span(qregs, check_node)
-
-                    if any(i in gate_span for i in all_indices):
-                        # needs to be a new layer
-                        layers.append([node])
-                    else:
-                        # can be added
-                        current_layer.append(node)
-                else:
-                    current_layer.append(node)
-
-            if current_layer:
-                layers.append(current_layer)
-            ops += layers
+                spooler.add(node, current_index)
+
+        ops = spooler.as_list()
 
     if justify == 'right':
+        spooler = LayerSpooler(qregs, Justification.RIGHT)
+
         dag_layers = []
 
         for dag_layer in dag.layers():
             dag_layers.append(dag_layer)
 
-        # Have to work from the end of the circuit
+        # going right to left!
         dag_layers.reverse()
 
-        # Dict per layer, keys are qubits and values are the gate
-        layer_dicts = [{}]
-
         for dag_layer in dag_layers:
+            current_index = 0
```

Comment:
If this is static could it not be declared in the `spooler.add()` method?

Reply:
Code has been refactored.